src/xaa.ts

```diff
@@ -142,14 +154,9 @@ export class TimeoutRunner {
    * @param tasks - Promise or function that returns Promise, or array of them.
    * @returns Promise to wait for tasks to complete, or timeout error.
    */
-  async run(tasks: Promise<any> | Function | Array<Function | Promise<any>>): Promise<any> {
-    const process = (x: any) => (typeof x === "function" ? x() : x);
-    let arrTasks: Promise<any[]>;
-    if (!Array.isArray(tasks)) {
-      arrTasks = process(tasks);
-    } else {
-      arrTasks = Promise.all(tasks.map(process));
-    }
+  async run(tasks: Runnable<T>) {
```

Comment:
Unfortunately this function still amounts to typing everything as `any`. `process` has a signature of `(x: any) => Promise<any>` because the type of `x` can't be inferred. That results in `arrTasks` being typed as `Promise<any>`, and finally `r` as `any`. Ultimately this results in `run` having a return type of `any`, too (because of issues like this, I'm a firm believer that at the least public functions should not rely on type inference for the return type) . `noImplicitAny` catches this error, we should consider turning it on.

I know we spent a lot of time typing this function without requiring `result` to be `T | T[]` but I don't think there's anyway around that. The reason being is: we have no idea if `run` will be called with a single `Task` or an array of `Tasks` when we create the `TimeoutRunner` instance. To put it another way, think about what the value of `result` will be in the plain JS version of this class:

```JavaScript
const t1 = new TimeoutRunner(2000, 'Too slow!');
// right now, t1.result is undefined

t1.run(async () => 'hello!');
// now t1.result is the value 'hello!'

const t2 = new TimeoutRunner(2000, 'Too slow!');
// right now, t2.result is undefined

t2.run([async () => 'hello!', async () => 'again!]);
// now t2.result is the value ['hello!', 'again!']
```

This shows the runtime behavior of this class and the behavior we have to capture in the type system. `result` is `T` or `T[]`, and there's no way to know which until `run` is called.

I think we should change `run` to the following:

```TypeScript
async run(tasks: Task<T>): Promise<T>;
async run(tasks: Array<Task<T>>): Promise<T[]>;
async run(tasks: Task<T> | Array<Task<T>>): Promise<T | T[]> {
  const process = async (x: Task<T>) => typeof x === "function" ? x() : x;
  const arrTasks = !Array.isArray(tasks) ? process(tasks) : Promise.all(tasks.map(process));
  try {
    const r = await Promise.race([arrTasks, this.defer.promise]);
    this.clear();
    this.result = r;
    return r;
  } catch (err) {
    this.clear();
    this.error = err;
    throw err;
  }
}
```

By using a function overload signature we can capture the correct return type based on what's passed in.

```TypeScript
const t3 = new TimeoutRunner<number>(100, '');

a.run(async() => 1); // Promise<number>
a.run([async() => 1]); // Promise<number[]>
```

Reply:
> I know we spent a lot of time typing this function without requiring result to be T | T[] but I don't think there's anyway around that. The reason being is: we have no idea if run will be called with a single Task or an array of Tasks when we create the TimeoutRunner instance. 

This is on point.

I tried to change run to the signature you suggested, but I couldn't get the implementation of `runTimeout` to work without a cast.  Can you check to see if my latest type signature is also sound?